targets/psa_switch/externs/psa_checksum.h

```diff
@@ -0,0 +1,41 @@
+
+
+#ifndef PSA_SWITCH_PSA_CHECKSUM_H_
+#define PSA_SWITCH_PSA_CHECKSUM_H_
+
+#include <bm/bm_sim/extern.h>
+#include <bm/bm_sim/calculations.h>
+
+
+namespace bm {
+
+namespace psa {
+
+class PSA_Checksum : public bm::ExternType {
+ public:
+  static constexpr p4object_id_t spec_id = 0xfffffffd;
+
+  BM_EXTERN_ATTRIBUTES {
+    BM_EXTERN_ATTRIBUTE_ADD(hash);
+  }
+
+  void init() override;
+
+  void get(Field& dst);
+
+  void get_verify(Field& dst, Field& equOp);
+  
+  void clear();
+
+  void update(const NamedCalculation& calculation);
```

Comment:
The PSA `update` extern method takes data to add to the checksum (incremental checksum computation), but this takes a `NamedCalculation` object which also includes a hash algorithm. It seems that the hash algorithm is a property of the extern instance, not of the calculation. Could you comment on the implementation choices in the code (there isn't much info to that effect in the commit message)?

Reply:
In p4c actual `update` parameters are encapsulated in calculation instance. So I made a calculation instance in JSON (in calculations block) that have information about update parameters (input field in calculation JSON) and about hash algorithm that this checksum instance use. So this parameter NamedCalculation is actually that parameter from JSON and it has information about hash algorithm and actual update parameters. It use hash algorithm which is defined in Checksum constructor on actual upadate parameters and calculate the checksum. I implemented like this because I thought it would be nice to use something that is already in code. BM project already has Calculation classes (NamedCalculation) where are hash algorithms already implemented and it has methods that are only used for calculating checksums. I didn't want to write something that already exists.

Reply:
When we use something like this in P4 code parsers:
	`ck.update({ parsed_hdr.ipv4.version,
                parsed_hdr.ipv4.ihl,
                parsed_hdr.ipv4.diffserv,
                parsed_hdr.ipv4.totalLen,
                parsed_hdr.ipv4.identification,
                parsed_hdr.ipv4.flags,
                parsed_hdr.ipv4.fragOffset,
                parsed_hdr.ipv4.ttl,
                parsed_hdr.ipv4.protocol,
                parsed_hdr.ipv4.srcAddr,
                parsed_hdr.ipv4.dstAddr });`
The JSON representation i `parsers` block is:
                  `"op" : "_Checksum_update",
                  "parameters" : [
                    {
                      "type" : "extern",
                      "value" : "IngressParserImpl.ck"
                    },
                    {
                      "type" : "calculation",
                      "value" : "calc"
                    }
                  ]`
Where calculation `calc` is reference to:
`{
      "name" : "calc",
      "id" : 0,
      "algo" : "crc16",
      "input" : [
        {
          "type" : "field",
          "value" : ["ipv4", "version"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ihl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "diffserv"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "totalLen"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "identification"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "flags"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "fragOffset"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ttl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "protocol"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "srcAddr"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "dstAddr"]
        }
      ]
    }`
In `calculations` JSON block.

Reply:
Thanks for all the detailed information. My only concern was that the hash function should ideally be a property of the extern. However, I understand that in the absence of a `field_list` construct in the bmv2 JSON, it is convenient to use a `calculation` for this. I think this is fine as an initial implementation, which can be refined later if we do introduce a `field_list` construct.

Reply:
although thinking about this a bit more (and remembering why I was opposed to adding a `field_list` construct to bmv2 in the first place), it seems that using a `parameters_vector` of integral values (search for `parameters_vector` in https://github.com/p4lang/behavioral-model/blob/main/docs/JSON_format.md) would have made more sense. I won't block the PR for this since these changes are limited to the PSA target, but I think a `parameters_vector` is the right approach as: 1) it may be applicable to other extern types as well, 2) `NamedCalculation` are basically legacy objects that predate P4_16, PSA and p4c.

Reply:
`Is this implementation with NamedCalculation good enough for merging?` I understand the solution with `parameters_vector`. But there are a few problems with that and that solution is not so straightforward. First hash algorithms and other structures that are used for checksum calculations are implemented in `calculations.cpp` in anonymous namespace so it is not available from other files. So, it is necessary to implement again the same code in `psa_checksum.cpp` file, or to change anonymous namespace (I don't know if is it ok to change it.). Second, if we use `paramameters_vector` in JSON, `PSA_Checksum::update` method need to look like this: `update(const std::vector<Data> fileds) `. JSON will look like: 
![parameters_vector](https://user-images.githubusercontent.com/77269536/122035238-8f8ada00-cdd2-11eb-89f8-73132263a11a.png)
So when we have package like this (`IPv4`): 4(version)5(ihl)00(diffserv)0038(totallen)...(in hex format), the elements of vector `fields` (which represent those fields `version`, `ihl` and so on) will have values 4,5,0,56 (decimal format). I think that there is no way to find the actual size of fields in `update` method, for example `diffserv`. Here it has value 0 which is 1B and actual size is 2B. Because of this, there is no way in method `update `to represent those fields as one single data to calculate checksum. This problem is solved with `Calculation` because when `calculations` are parsed in JSON it use BufBuilder and other structures to resolve this problem (BufBuilder saves info about header, offset).  When `parameters_vector` is parsed in JSON it only takes actual fields values and represents them as `Data`. So, in method `update` with vector<Data> I only have actual values. There is no information about actual field (size, header, field name etc.).

Reply:
Also, `PSA_Checksum::update` shouldn't have vector of fields as parameter because when we use construct `{field1,field2,field3,...}` in `update` in P4, it represents kind of concatenation and those fields are packed in one object. It would be necessary to introduce new JSON construct which would `tie` all fields from `update` JSON in one `Data` in bm. This has already been done with `calculations`. `Calculations` construct was made because of this problem. Only thing that is problematic with `calculation` implementation is that `PSA_Checksum` doesn't have information about `hash` as attribute. But it is not really necessary. Hash is only important for `update`, so in p4c `hash` attribute is taken from extern checksum constructor and given to corresponding calculation to calculate checksum.

Reply:
I agree with your first point (missing size information is an issue). I am not sure I agree with the second statement though. Ultimately, it's all serialized as one single byte buffer, the structure of the input doesn't matter. It's a valid compiler transformation IMO to flatten the input so it's just a sequence of fields, and only the order matters.

Reply:
I have only tried to say that it is important to save actual sizes of fields in package, so the actual value from package could be recreated in `update` or to make new JSON construct that would "tie" the fields in one single data while parsing JSON in `P4Objects`. Parameter `std::vector<Data> fields` doesn't support that.
Maybe the checksum computation in PSA should be done with `Hash` extern, which is used to calculate hashes in PSA. For example `PSA_Checksum` should have extern `Hash` object as attribute. `PSA_Chekcsum` attribute `hash` which is a `string` should be used as `Hash` object's constructor parameter in `PSA_Checksum::init` where it should be initialized. Then in `PSA_Checksum::update`, the extern `Hash` object will be used for calculating checksum over the package. But extern `Hash` isn't supported yet in PSA.
This pull request is about issue number #974 and it fix it. Now user can use extern checksum and its methods to calculate checksum. This implementation has its weakness (hash isn't `PSA_Checksum` attribute in bm) but for actual PSA architecture that is supported, I think this implementation with `NamedCalculation` as `PSA_Checksum::update` parameter is the best approach.

Reply:
I made new implementation for `PSA_Checksum::update` as you requested @antoninbas. Now, PSA_Checksum::update method has  declaration: `update(const std::vector<Field> fields)`. 
JSON looks like: 
![fields_list](https://user-images.githubusercontent.com/77269536/122771475-bfd8e980-d2a6-11eb-86be-0b51cf3728fb.png)
PSA Checksum JSON instance has information about its `hash` attribute in constructor.
`PSA_Checksum` class in bm has the `hash`attribute. In method `update`, checksum will be calculated based on which hash the current checksum instance uses. The hash implementations are used from `calculations.cpp`.